collector.go

```diff
@@ -81,3 +69,113 @@ func StartBundleCollector(merkleCh chan<- []MerkleRootEntry, errorCh chan<- erro
 		paginationKey = nextKey
 	}
 }
+
+// the bundle collector downloads and unpacks the bundles data,
+// calculates the merkle hash and puts it into the merkle entry channel
+func StartBundleCollector(context context.Context, merkleEntries chan<- MerkleRootEntry, bundleCh <-chan BundleInfo, errorCh chan<- error, storageRest string) {
+	for {
+
+		select {
+		case <-context.Done():
+			return
+		case bundle := <-bundleCh:
+			decompressedBundle, err := bundles.GetDataFromFinalizedBundle(bundle.Bundle, storageRest)
+
+			if err != nil {
+				logger.Info().
+					Str("bundleId", bundle.Bundle.Id).
+					Int("poolId", bundle.PoolId).
+					Msg("error while fetching")
+				errorCh <- err
+				return
+			}
+
+			// parse bundle
+			var dataItems []types.DataItem
+
+			if err := json.Unmarshal(decompressedBundle, &dataItems); err != nil {
+				errorCh <- fmt.Errorf("failed to unmarshal tendermint bundle: %w", err)
+				return
+			}
+
+			leafHashes := BundleToHashes(dataItems, bundle.Runtime)
+			merkleRoot := GenerateMerkleRoot(&leafHashes)
+
+			merkleEntry := MerkleRootEntry{BundleId: bundle.BundleId, Hash: merkleRoot, PoolId: bundle.PoolId}
+			select {
+			case <-context.Done():
+				return
+			case merkleEntries <- merkleEntry:
+				logger.Info().
+					Int("bundleId", merkleEntry.BundleId).
+					Int("poolId", merkleEntry.PoolId).
+					Str("root", hex.EncodeToString(merkleEntry.Hash[:])).
+					Msg("computed Merkle root")
+				continue
+			}
+		}
+	}
+}
+
+// the merkle writer collects all the merkle entries for each pool and appends them in order, so that no merkle entry is missing
+func StartMerkleWriter(context context.Context, merkleEntries <-chan MerkleRootEntry, errorCh chan<- error, pools []Pool, cancel context.CancelFunc, poolHeights map[int]int) {
+
+	poolEntries := map[int]*MerkleRootQueue{}
+	if reachedTargetHeight(poolHeights, pools) {
+		cancel()
+		return
+	}
+
+	for {
+		select {
+		case <-context.Done():
+			return
+		case entry := <-merkleEntries:
+			// insert the entry into the priority queue
+			queue := poolEntries[entry.PoolId]
+			if queue == nil {
+				queue = &MerkleRootQueue{}
+				poolEntries[entry.PoolId] = queue
+			}
+			heap.Push(queue, entry)
+			// write all hashes that are in order and come next
+			for queue.Len() > 0 && (*queue)[0].BundleId == poolHeights[entry.PoolId] {
+				entry := heap.Pop(queue).(MerkleRootEntry)
+				poolHeights[entry.PoolId]++
+				appendMerkleRoot(entry)
+				logger.Info().Int("height", poolHeights[entry.PoolId]).Int("pool", entry.PoolId).Msg("writing hashes")
```

Comment:
Use `bundle-id` instead of height and `writing merkle-root` instead of `writing hashes`